test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go

```diff
@@ -0,0 +1,200 @@
+
+	gcm "google.golang.org/api/monitoring/v3"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/discovery"
+	kubeaggrcs "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+	customclient "k8s.io/metrics/pkg/client/custom_metrics"
+)
+
+var _ = instrumentation.SIGDescribe("Stackdriver Monitoring", func() {
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
+	})
+
+	f := framework.NewDefaultFramework("stackdriver-monitoring")
+	var kubeClient clientset.Interface
+	var kubeAggrClient kubeaggrcs.Interface
+	var customMetricsClient customclient.CustomMetricsClient
+	var discoveryClient *discovery.DiscoveryClient
+
+	It("should run Custom Metrics - Stackdriver Adapter [Feature:StackdriverMonitoring]", func() {
+		kubeClient = f.ClientSet
+		kubeAggrClient = f.AggregatorClient
+		config, err := framework.LoadConfig()
+		if err != nil {
+			framework.Failf("Failed to load config: %s", err)
+		}
+		customMetricsClient = customclient.NewForConfigOrDie(config)
+		discoveryClient = discovery.NewDiscoveryClientForConfigOrDie(config)
+		testAdapter(f, kubeClient, customMetricsClient, discoveryClient)
+	})
+})
+
+func testAdapter(f *framework.Framework, kubeClient clientset.Interface, customMetricsClient customclient.CustomMetricsClient, discoveryClient *discovery.DiscoveryClient) {
+	projectId := framework.TestContext.CloudConfig.ProjectID
+
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, gcm.CloudPlatformScope)
+
+	// Hack for running tests locally
+	// If this is your use case, create application default credentials:
+	// $ gcloud auth application-default login
+	// and uncomment following lines (comment out the two lines above):
+	/*
+		ts, err := google.DefaultTokenSource(oauth2.NoContext)
+		framework.Logf("Couldn't get application default credentials, %v", err)
+		if err != nil {
+			framework.Failf("Error accessing application default credentials, %v", err)
+		}
+		client := oauth2.NewClient(oauth2.NoContext, ts)
+	*/
+
+	gcmService, err := gcm.New(client)
+	if err != nil {
+		framework.Failf("Failed to create gcm service, %v", err)
+	}
+
+	framework.ExpectNoError(err)
+
+	// Set up a cluster: create a custom metric and set up k8s-sd adapter
+	err = createDescriptors(gcmService, projectId)
+	if err != nil {
+		framework.Failf("Failed to create metric descriptor: %s", err)
+	}
+	defer cleanupDescriptors(gcmService, projectId)
+
+	err = createAdapter()
+	if err != nil {
+		framework.Failf("Failed to set up: %s", err)
+	}
+	defer cleanupAdapter()
+
+	// Run application that exports the metric
+	err = createSDExporterPod(kubeClient)
+	if err != nil {
+		framework.Failf("Failed to create sd-exporter pod: %s", err)
+	}
+	defer cleanupSDExporterPod(kubeClient)
+
+	// Wait a short amount of time to create a pod and export some metrics
+	time.Sleep(60 * time.Second)
```

Comment:
I think you should explicitly wait for pod to be schedule or metric to have some values in SD. Otherwise this test might be very flaky. WDYT?

Reply:
If you agree we might fix it in the follow-up PR.

Reply:
I agree, this also seems to be a common pattern in integration tests.